pkg/csi/service/vanilla/controller.go

```diff
@@ -764,11 +764,28 @@ func (c *controller) ControllerPublishVolume(ctx context.Context, req *csi.Contr
 				},
 			}
 			// Select only the backing object details.
-			queryResult, err := c.manager.VolumeManager.QueryAllVolume(ctx, queryFilter, querySelection)
-			if err != nil {
-				msg := fmt.Sprintf("QueryVolume failed for volumeID: %q. %+v", req.VolumeId, err.Error())
-				log.Error(msg)
-				return nil, status.Error(codes.Internal, msg)
+			var queryResult *cnstypes.CnsQueryResult
+			var queryAsyncNotSupported bool
+			if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.AsyncQueryVolume) {
+				// AsyncQueryVolume feature switch is disabled
+				queryResult, err = c.manager.VolumeManager.QueryVolumeAsync(ctx, queryFilter, querySelection)
+				if err.Error() == cnsvsphere.ErrNotSupported.Error() {
+					log.Warn("QueryVolumeAsync is not supported. Invoking QueryVolume API")
+					queryAsyncNotSupported = true
+				} else {
+					msg := fmt.Sprintf("QueryVolumeAsync failed for volumeID: %q. %+v", req.VolumeId, err.Error())
+					log.Error(msg)
+					return nil, status.Error(codes.Internal, msg)
+				}
+			}
+			// If UseQueryAsync feature state is disabled or QueryAsync is not supported by the current vSphere version
+			if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.AsyncQueryVolume) || queryAsyncNotSupported {
```

Comment:
Shouldnt this be 
```
if !commonco.ContainerOrchestratorUtility.IsFSSEnabled( ...
```

Reply:
My bad. Must have missed this. Thanks for catching this one!